### PR TITLE
Dra 1203 enrich metadata

### PR DESCRIPTION
**Pull request by davidgrove73**

### CODE REVIEW

File: src/main/java/dk/kb/datahandler/enrichment/DataEnricher.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    //    record.setData(enrichedData);
+        return record;
+    }
+
+    private static Document addMetadataFragments(Document record, Document fragments) {
+        // TODO check if fragment exists
+        NodeList nodeList = record.getElementsByTagName("XIP");
+        if (nodeList.getLength() > 0 ) {
+            Node xipNode = nodeList.item(0);
+            Element metadataNode = record.createElement("Metadata");
```
Comment (VictorHarbo) [inline src/main/java/dk/kb/datahandler/enrichment/DataEnricher.java:59]: I think this element should have the schemaUri from the metadataContainer in the fragment eg. `schemaUri="http://id.kb.dk/schemas/supplementary_tvmeter_metadata"`

File: src/main/java/dk/kb/datahandler/enrichment/FragmentsClient.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+public class FragmentsClient {
+
+    private static FragmentsClient instance;
+
+    private String baseUrl;
+
+    public static synchronized FragmentsClient getInstance() {
+        if (instance == null) {
+            instance = new FragmentsClient(ServiceConfig.getConfig().getString("fragmentService.baseUrl"));
+        }
```
Comment (VictorHarbo) [inline src/main/java/dk/kb/datahandler/enrichment/FragmentsClient.java:24]: Remember to add this value to aegis at some point :)

File: src/main/java/dk/kb/datahandler/enrichment/FragmentsClient.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            //do some error handling
+            return null;
+        }
+        BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
+        String inputLine;
+        StringBuilder response = new StringBuilder();
+        while ((inputLine = in.readLine()) != null) {
+            response.append(inputLine);
+        }
+        in.close();
```
Comment (VictorHarbo) [inline src/main/java/dk/kb/datahandler/enrichment/FragmentsClient.java:45]: These lines are duplicated in the DsPreservica client. Should we implement a client utils maybe?

**Review by VictorHarbo — CHANGES_REQUESTED** [review]
Some comments on the code and a couple of documentation comments, which i might have more of, when this isn't a draft anymore. 
Thomas has returned from vacation and has come up with a good argument for making a change in how we implement this, we should discuss this all four of us.

**Comment by jorntx** [comment]
Very good suggestions @VictorHarbo . @davidgrove73 I looked at it too and think it provides a very good start. We have to take it in a slightly other direction due to the fact that we only want "complete/valid (enriched)" records in storage to prevent runtime errors. The base of the code is however reusable. More on this topic later:)

File: src/main/java/dk/kb/datahandler/enrichment/DataEnricher.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import org.w3c.dom.*;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.TransformerException;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.List;
+
+public class DataEnricher {
```
[the remark this answers is not in the document]
Comment (davidgrove73) [reply src/main/java/dk/kb/datahandler/enrichment/DataEnricher.java:21]: java doc added

File: src/main/java/dk/kb/datahandler/enrichment/DataEnricher.java (hunk 1 above)
Comment (davidgrove73) [reply src/main/java/dk/kb/datahandler/enrichment/DataEnricher.java:72]: all attributes from the metadataContainer element in the fragment is copied to the Metadata element (see lines 75-80)

File: src/test/java/dk/kb/datahandler/enrichment/FragmentClientTest.java
[the remark this answers is not in the document]
Comment (davidgrove73) [reply src/test/java/dk/kb/datahandler/enrichment/FragmentClientTest.java:1]: Some tests has been added :-)

File: src/main/java/dk/kb/datahandler/enrichment/FragmentsClient.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    private static final Logger log = LoggerFactory.getLogger(FragmentsClient.class);
+
+    private static FragmentsClient instance;
+
+    private final String baseUrl;
+    private final int maxRetries;
+
+    public static synchronized FragmentsClient getInstance() {
+        if (instance == null) {
+            instance = new FragmentsClient(ServiceConfig.getConfig().getString("fragmentService.baseUrl"),5);
```
Comment (VictorHarbo) [inline src/main/java/dk/kb/datahandler/enrichment/FragmentsClient.java:30]: I believe five retries is a fine number. However, should we consider making this configurable? I don't have a strong opinion on this and one could argue that this is a one time thing, which probably shouldn't be configured anyway. Lets keep it at 5 hardcoded.

File: src/test/java/dk/kb/datahandler/enrichment/FragmentClientTest.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+
+
+    @Test
+    public void testFetchFragments() throws IOException, URISyntaxException {
+        when(mockConnection.getResponseCode()).thenReturn(200);
+
+        String jsonResponse = Files.readString(Resolver.getPathFromClasspath("xml/fragments-multi.json"));
+        InputStream jsonStream = new java.io.ByteArrayInputStream(jsonResponse.getBytes());
```
Comment (VictorHarbo) [inline src/test/java/dk/kb/datahandler/enrichment/FragmentClientTest.java:42]: This line throws a Forbidden API error, please specifiy charset

File: src/test/java/dk/kb/datahandler/enrichment/FragmentClientTest.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Test
+    public void testRetriesOnFailure() throws IOException, URISyntaxException {
+        when(mockConnection.getResponseCode())
+                .thenReturn(500)
+                .thenReturn(500)
+                .thenReturn(200);
+
+        when(mockConnection.getInputStream()).thenReturn(new java.io.ByteArrayInputStream("[]".getBytes()));
```
Comment (VictorHarbo) [inline src/test/java/dk/kb/datahandler/enrichment/FragmentClientTest.java:56]: This line also throws a Forbidden API error, please specify charset

File: src/main/openapi/ds-datahandler-openapi_v1.yaml [hunk 8]
```diff
@@ -428,8 +426,11 @@ components:
             * `date`:     Only use format yyyy-MM-dd as from/until parameters
             * `datetime`:    Use UTC format yyyy-MM-ddThh:mm:ssZ as from/until parameters. This is default.                        
           example: 'date'
-          default: 'date'         
-          
+          default: 'date'
+        fragmentServiceUrl:
+          type: string
+          description: BaseURL for the enrichment fragment service
+          example: 'http://example.com/'
```
Comment (VictorHarbo) [inline src/main/openapi/ds-datahandler-openapi_v1.yaml:433]: This is indeed the correct way. Could you add an example value for the origin ds.radiotv in ds-datahandler-behaviour.yaml oaiTargets and also in the corresponding aegis file for the devel system?

File: src/test/java/dk/kb/datahandler/enrichment/DataEnricherTest.java
Comment (VictorHarbo) [inline src/test/java/dk/kb/datahandler/enrichment/DataEnricherTest.java:1]: When i run these tests together in IntelliJ IDEA I get the following error: 

org.mockito.exceptions.base.MockitoException: 
For dk.kb.datahandler.enrichment.FragmentsClient, static mocking is already registered in the current thread

To create a new mock, the existing static mock registration must be deregistered

File: src/test/java/dk/kb/datahandler/enrichment/DataEnricherTest.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+import static org.mockito.Mockito.when;
+
+public class DataEnricherTest {
+
+    private FragmentsClient fragmentsClientMock;
+    private ObjectMapper mapper;
+
+    @BeforeEach
+    public void setUp() throws IOException, URISyntaxException {
+        ServiceConfig.initialize("conf/ds-datahandler-behaviour.yaml","conf/ds-datahandler-local.yaml");
```
Comment (VictorHarbo) [inline src/test/java/dk/kb/datahandler/enrichment/DataEnricherTest.java:37]: My guess is that you've made some configuration changes in your local config. For testing purposes could these be moved to aegis in the file: `ds-datahandler/local/src/test/resources/ds-datahandler-integration-test.yaml` or another test-config file placed in aegis. To see how this has been implemted have a look at the `setup`-method in datahandlers `OaiHarvestClientIntegrationTest`-class.

**Review by VictorHarbo — CHANGES_REQUESTED** [review]
Some comments on tests and test setup and a small typo otherwise we are going in a good direction here.

File: src/main/java/dk/kb/datahandler/enrichment/DataEnricher.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            for (int i = 0; i < childNodes.getLength(); i++) {
+                Node child = record.importNode(childNodes.item(i), true);
+                metadataNode.appendChild(child);
+            }
+            xipNode.appendChild(metadataNode);
+
+        }
+    }
+
+    private static String extractOiId(String recordId) {
```
[the remark this answers is not in the document]
Comment (davidgrove73) [reply src/main/java/dk/kb/datahandler/enrichment/DataEnricher.java]: fixed

File: src/test/java/dk/kb/datahandler/enrichment/FragmentClientTest.java (hunk 6 above)
Comment (davidgrove73) [reply src/test/java/dk/kb/datahandler/enrichment/FragmentClientTest.java]: should be fixed

File: src/test/java/dk/kb/datahandler/enrichment/FragmentClientTest.java (hunk 7 above)
Comment (davidgrove73) [reply src/test/java/dk/kb/datahandler/enrichment/FragmentClientTest.java]: should be fixed

File: src/main/openapi/ds-datahandler-openapi_v1.yaml (hunk 8 above)
Comment (davidgrove73) [reply src/main/openapi/ds-datahandler-openapi_v1.yaml:433]: done

File: src/test/java/dk/kb/datahandler/enrichment/DataEnricherTest.java
Comment (davidgrove73) [reply src/test/java/dk/kb/datahandler/enrichment/DataEnricherTest.java:1]: fixed

File: src/main/openapi/ds-datahandler-openapi_v1.yaml (hunk 8 above)
Comment (VictorHarbo) [reply src/main/openapi/ds-datahandler-openapi_v1.yaml:433]: I can't find the updated version in aegis for the devel system, but I like how its done in the behaviour file!

File: src/test/java/dk/kb/datahandler/enrichment/DataEnricherTest.java (hunk 9 above)
Comment (VictorHarbo) [reply src/test/java/dk/kb/datahandler/enrichment/DataEnricherTest.java]: I see that you've removed all configuration from the test class. That's pretty!

**Comment by VictorHarbo** [comment]
I've review during the draft process. Rebasing and merging now.

**Pull request merged**